geo/src/test/java/org/odk/collect/geo/geopoly/GeoPolyActivityTest.kt

```diff
@@ -0,0 +1,291 @@
+        assertThat(polys[0].points.isEmpty(), equalTo(true))
+    }
+
+    @Test
+    fun whenPolygonExtraPresent_andPolyIsEmpty_pressingBack_finishes() {
+        val intent =
+            Intent(ApplicationProvider.getApplicationContext(), GeoPolyActivity::class.java)
+        val polygon = ArrayList<MapPoint>()
+        intent.putExtra(GeoPolyActivity.EXTRA_POLYGON, polygon)
+        intent.putExtra(GeoPolyActivity.OUTPUT_MODE_KEY, GeoPolyActivity.OutputMode.GEOSHAPE)
+        val scenario = launcherRule.launch<GeoPolyActivity>(intent)
+        mapFragment.ready()
+        Espresso.pressBack()
+        assertThat(scenario.isFinishing, equalTo(true))
+    }
+
+    @Test
+    fun startingInput_usingAutomaticMode_usesRetainMockAccuracyTrueToStartLocationTracker() {
+        val intent =
+            Intent(ApplicationProvider.getApplicationContext(), GeoPolyActivity::class.java)
+        intent.putExtra(Constants.EXTRA_RETAIN_MOCK_ACCURACY, true)
+        launcherRule.launch<GeoPolyActivity>(intent)
+        mapFragment.ready()
+        startInput(R.id.automatic_mode)
+        verify(locationTracker).start(true)
+    }
+
+    @Test
+    fun startingInput_usingAutomaticMode_usesRetainMockAccuracyFalseToStartLocationTracker() {
+        val intent =
+            Intent(ApplicationProvider.getApplicationContext(), GeoPolyActivity::class.java)
+        intent.putExtra(Constants.EXTRA_RETAIN_MOCK_ACCURACY, false)
+        launcherRule.launch<GeoPolyActivity>(intent)
+        mapFragment.ready()
+        startInput(R.id.automatic_mode)
+        verify(locationTracker).start(false)
+    }
+
+    @Test
+    fun recordingPointManually_whenPointIsADuplicateOfTheLastPoint_skipsPoint() {
+        launcherRule.launch(GeoPolyActivity::class.java)
+        mapFragment.ready()
+        startInput(R.id.manual_mode)
+        mapFragment.setLocation(MapPoint(1.0, 1.0))
+        onView(withId(R.id.record_button)).perform(click())
+        onView(withId(R.id.record_button)).perform(click())
+        assertThat(mapFragment.getPolyLines()[0].points.size, equalTo(1))
+    }
+
+    @Test
+    fun placingPoint_whenPointIsADuplicateOfTheLastPoint_skipsPoint() {
+        launcherRule.launch(GeoPolyActivity::class.java)
+        mapFragment.ready()
+        startInput(R.id.placement_mode)
+        mapFragment.click(MapPoint(1.0, 1.0))
+        mapFragment.click(MapPoint(1.0, 1.0))
+        assertThat(mapFragment.getPolyLines()[0].points.size, equalTo(1))
+    }
+
+    @Test
+    fun buttonsShouldBeEnabledInEditableMode() {
+        val polyline = ArrayList<MapPoint>()
+        polyline.add(MapPoint(1.0, 2.0, 3.0, 4.0))
+        val intent =
+            Intent(ApplicationProvider.getApplicationContext(), GeoPolyActivity::class.java)
+        intent.putExtra(GeoPolyActivity.EXTRA_POLYGON, polyline)
+        val scenario = launcherRule.launch<GeoPolyActivity>(intent)
+        mapFragment.ready()
+        scenario.onActivity { activity: GeoPolyActivity ->
```

Comment:
All these `onActivity` calls could be combined. Same for the below test.

Reply:
Fixed.